Use FunctionKeyword for computed_value of timing function

hiikezoe opened the pull request:
<!-- Please describe your changes on the following line: -->
https://bugzilla.mozilla.org/show_bug.cgi?id=1358330

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #15086

<!-- Either: -->
- [X] These changes do not require tests because mozilla-central has test cases.  There might be some test cases in servo tree.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16640)
<!-- Reviewable:end -->

Commit:
Factor out serialize keywoded timing function values.

This will be used for computed timing function as well.

Commit:
Use FunctionKeyword for computed_value of timing-function as well.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/longhand/box.mako.rs, components/style/animation.rs, components/style/gecko_bindings/sugar/ns_timing_function.rs
 * @emilio: components/style/properties/longhand/box.mako.rs, components/style/animation.rs, components/style/gecko_bindings/sugar/ns_timing_function.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style code, but no tests are modified. Please consider adding a test!

nox commented on components/style/animation.rs:
```diff
@@ -367,7 +372,10 @@ impl PropertyAnimation {
-            }
+            },
+            TransitionTimingFunction::Keyword(_) => {
+                unreachable!("Keyword function should not appear")
```
You probably want `panic!` here, `unreachable!` AFAIK is for more obvious code, not for invariants maintained elsewhere.

hiikezoe replied:
Thank you!

mbrubeck commented:
@bors-servo delegate+

bors-servo commented:
:v: @hiikezoe can now approve this pull request

hiikezoe commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit fc95bee0e6d7a78407e10a14b845859859ba5d45 with merge f1820a86e3be773e22e4e8f3b9ecc00a54d5a31e...

bors-servo commented:
:broken_heart: Test failed - [arm32](http://build.servo.org/builders/arm32/builds/6818)

hiikezoe commented:
@bors-servo try

hiikezoe commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit f15896078ac136cdc97ac5a8678d5e2e949cab8f with merge 353556d7063d84d4c20a076a3887118b84083d1f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3697)

hiikezoe commented:
OK. All of failure tests for timing-function are wrong. 
Those are also marked as 'FAIL' on mozilla-central. Sigh.

https://hg.mozilla.org/mozilla-central/file/2cca333f546f/testing/web-platform/meta/css/css-transitions-1/transition-001.html.ini
https://hg.mozilla.org/mozilla-central/file/2cca333f546f/testing/web-platform/meta/css/css-transitions-1/transition-timing-function-001.html.ini

hiikezoe commented:
I'd like to fix the test cases in mozilla-central because it will be synced with upstream.
Filed; https://bugzilla.mozilla.org/show_bug.cgi?id=1360410

Commit:
Mark some tests for timing-function as FAIL.

Those test cases are apparently wrong. Keyword values (e.g. 'ease')
should not be serialized as cubic-bezier function.
https://drafts.csswg.org/css-timing/#serializing-a-timing-function

On gecko, we also mark those tests FAIL.

https://hg.mozilla.org/mozilla-central/file/2cca333f546f/testing/web-platform/meta/css/css-transitions-1/transition-001.html.ini
https://hg.mozilla.org/mozilla-central/file/2cca333f546f/testing/web-platform/meta/css/css-transitions-1/transition-timing-function-001.html.ini

hiikezoe commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 24b808a87ad38ee98d78945b3d7c62b3b0547516 with merge 276034be983ebcd1716d0be485e7467080449040...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3699)

jdm commented:
@hiikezoe Servo's web-platform-tests also syncs with upstream.

hiikezoe commented:
Thank you @jdm.  Xidorn also notice me that we should not touch tests added by someone else, so I leave the failure tests as they are here.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit ed4ae685897f488f80b0bb2dc5185c046b9e749e with merge daca90f8b9565007f902f39160e711a828132f42...

jdm commented:
There's nothing wrong with touching tests if we're actually confident that the changes are correct.

hiikezoe commented:
OK,  I will do change tests with PR next time. Thank you!

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3702)

hiikezoe commented:
Coping .ini files from mozilla-central were slightly different..
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b5bcc6e5ebc25fd9acd26f1cf05b9bf4dca75dd6 with merge 5f95d8058fcdbb2360a9dbd0f965f673fbb40797...

bors-servo commented:
:boom: Test timed out

nox commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit b5bcc6e5ebc25fd9acd26f1cf05b9bf4dca75dd6 with merge 23aaa0f6ec09f086d1ebf2ef76bf18a4f0c46fed...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3744)

hiikezoe commented:
@bors-servo r=xidorn

bors-servo commented:
:pushpin: Commit b5bcc6e has been approved by `xidorn`

<!-- @bors-servo r=xidorn b5bcc6e5ebc25fd9acd26f1cf05b9bf4dca75dd6 -->

bors-servo commented:
:hourglass: Testing commit b5bcc6e5ebc25fd9acd26f1cf05b9bf4dca75dd6 with merge 93a513c06fe6e889efc5a3a52a1509298a81bb3e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3747)

hiikezoe commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [arm64](http://build.servo.org/builders/arm64/builds/6872), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6983), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6983), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3672), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2170) are reusable. Rebuilding only [android](http://build.servo.org/builders/android/builds/6630), [arm32](http://build.servo.org/builders/arm32/builds/6843), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3721), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3747), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/7016), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3739)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6630), [arm32](http://build.servo.org/builders/arm32/builds/6843), [arm64](http://build.servo.org/builders/arm64/builds/6872), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6983), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3721), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3748), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6983), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/7016), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3672), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3739), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2170)
Approved by: xidorn
Pushing 93a513c06fe6e889efc5a3a52a1509298a81bb3e to master...
